File: dashboard/src/api/policies.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025-2026 The Phantex Authors

/**
 * Phantex — Policy API hooks (O5).
 *
 * TanStack Query hooks for the policy CRUD API:
 *   - usePolicies: list with pagination
 *   - usePolicy: single policy detail
 *   - useCreatePolicy / useUpdatePolicy / useDeletePolicy: mutations
 *   - useValidatePolicy: live YAML/JSON validation
 *   - useApplyPolicy: apply policy to agents
 *   - usePolicyVersions: version history
 *
 * All mutations invalidate the policy list cache on success.
 *
 * @module api/policies
 */

import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import apiClient from "@/api/client"
import type {
  Policy,
  PolicyDefinition,
  PolicyValidationResult,
  PolicyVersion,
  PaginatedResponse,
} from "@/types"

/* ── Query Keys ────────────────────────────────────────────────────────────── */

const KEYS = {
  all: ["policies"] as const,
  list: (page: number, pageSize: number, enabledOnly: boolean) =>
    ["policies", "list", page, pageSize, enabledOnly] as const,
  detail: (id: string) => ["policies", "detail", id] as const,
  versions: (id: string) => ["policies", "versions", id] as const,
}

/* ── Helpers ───────────────────────────────────────────────────────────────── */

/** Defence-in-depth: UUID-only path segment. */
function safePath(id: string): string {
  if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(id)) {
    throw new Error("Invalid policy ID format")
  }
  return encodeURIComponent(id)
}

/* ── List Policies ─────────────────────────────────────────────────────────── */

export function usePolicies(
  { page = 1, pageSize = 20, enabledOnly = false } = {},
  enabled = true,
) {
  return useQuery<PaginatedResponse<Policy>>({
    queryKey: KEYS.list(page, pageSize, enabledOnly),
    queryFn: async () => {
      const { data } = await apiClient.get("/policies", {
        params: { page, page_size: pageSize, enabled_only: enabledOnly },
      })
      return data
    },
    enabled,
    staleTime: 10_000,
  })
}

/* ── Single Policy ─────────────────────────────────────────────────────────── */

export function usePolicy(id: string | undefined, enabled = true) {
  return useQuery<Policy>({
    queryKey: KEYS.detail(id ?? ""),
    queryFn: async () => {
      const { data } = await apiClient.get(`/policies/${safePath(id!)}`)
      return data
    },
    enabled: !!id && enabled,
    staleTime: 15_000,
  })
}

/* ── Create Policy ─────────────────────────────────────────────────────────── */

interface CreatePolicyInput {
  name: string
  description?: string
  enabled?: boolean
  definition: PolicyDefinition
}

export function useCreatePolicy() {
  const qc = useQueryClient()
  return useMutation<Policy, Error, CreatePolicyInput>({
    mutationFn: async (input) => {
      const { data } = await apiClient.post("/policies", input)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: KEYS.all })
    },
  })
}

/* ── Update Policy ─────────────────────────────────────────────────────────── */

interface UpdatePolicyInput {
  id: string
  name?: string
  description?: string
  enabled?: boolean
  definition?: PolicyDefinition
  change_summary?: string
}

export function useUpdatePolicy() {
  const qc = useQueryClient()
  return useMutation<Policy, Error, UpdatePolicyInput>({
    mutationFn: async ({ id, ...body }) => {
      const { data } = await apiClient.put(`/policies/${safePath(id)}`, body)
      return data
    },
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({ queryKey: KEYS.all })
      qc.invalidateQueries({ queryKey: KEYS.detail(vars.id) })
      qc.invalidateQueries({ queryKey: KEYS.versions(vars.id) })
    },
  })
}

/* ── Delete Policy ─────────────────────────────────────────────────────────── */

export function useDeletePolicy() {
  const qc = useQueryClient()
  return useMutation<void, Error, string>({
    mutationFn: async (id) => {
      await apiClient.delete(`/policies/${safePath(id)}`)
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: KEYS.all })
    },
  })
}

/* ── Validate Policy ───────────────────────────────────────────────────────── */

interface ValidateInput {
  yaml_content?: string
  json_content?: PolicyDefinition
}

export function useValidatePolicy() {
  return useMutation<PolicyValidationResult, Error, ValidateInput>({
    mutationFn: async (input) => {
      const { data } = await apiClient.post("/policies/validate", input)
      return data
    },
  })
}

/* ── Apply Policy ──────────────────────────────────────────────────────────── */

interface ApplyResult {
  status: string
  policy_id: string
  policy_name: string
  rules_count: number
  scope: { agent_tags: string[]; frameworks: string[] }
}

export function useApplyPolicy() {
  return useMutation<ApplyResult, Error, string>({
    mutationFn: async (id) => {
      const { data } = await apiClient.post(`/policies/${safePath(id)}/apply`)
      return data
    },
  })
}

/* ── Policy Versions ───────────────────────────────────────────────────────── */

export function usePolicyVersions(policyId: string | undefined, enabled = true) {
  return useQuery<PolicyVersion[]>({
    queryKey: KEYS.versions(policyId ?? ""),
    queryFn: async () => {
      const { data } = await apiClient.get(
        `/policies/${safePath(policyId!)}/versions`,
      )
      return data
    },
    enabled: !!policyId && enabled,
    staleTime: 30_000,
  })
}

File: backend/app/routers/policies.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2025-2026 The Phantex Authors

"""
Phantex — Policy Editor Router.

CRUD API for detection policies with YAML support, versioning,
and validation.

All endpoints are tenant-scoped via auth. Policy CRUD requires
admin or analyst role.

Endpoints:
  GET    /api/v1/policies              — List policies
  POST   /api/v1/policies              — Create policy
  GET    /api/v1/policies/:id          — Get policy detail
  PUT    /api/v1/policies/:id          — Update policy
  DELETE /api/v1/policies/:id          — Soft-delete policy
  POST   /api/v1/policies/validate     — Validate YAML/JSON without saving
  POST   /api/v1/policies/:id/apply    — Apply policy to matching agents
  GET    /api/v1/policies/:id/versions — Get version history
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query

from app.database import get_raw_db
from app.middleware.auth import get_current_active_user
from app.middleware.rate_limit import rate_limit
from app.schemas.auth import CurrentUser
from app.schemas.policy import (
    PolicyCreateRequest,
    PolicyListResponse,
    PolicyResponse,
    PolicyUpdateRequest,
    PolicyValidateRequest,
    PolicyValidationResult,
    PolicyVersionResponse,
)
from app.services.policy_service import (
    create_policy,
    delete_policy,
    get_policy,
    get_policy_versions,
    list_policies,
    parse_yaml_safe,
    update_policy,
    validate_policy_definition,
)
from app.utils.logging import get_logger

logger = get_logger("phantex.router.policies")

router = APIRouter(
    prefix="/api/v1/policies",
    tags=["policies"],
    dependencies=[Depends(rate_limit)],
)

# ── Auth Helpers ──────────────────────────────────────────────────────────────

def _require_policy_role(user: CurrentUser) -> None:
    """Raise 403 if user doesn't have admin or analyst role."""
    if user.role not in ("admin", "analyst"):
        raise HTTPException(
            status_code=403,
            detail="Policy management requires admin or analyst role",
        )

# ── List Policies ────────────────────────────────────────────────────────────

@router.get("", response_model=PolicyListResponse)
async def list_policies_endpoint(
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    enabled_only: bool = Query(False),
):
    """List policies for the current tenant."""
    tenant_id = str(user.tenant_id)
    await conn.set_tenant(tenant_id)
    items, total = await list_policies(conn, tenant_id, page=page, page_size=page_size, enabled_only=enabled_only)
    return PolicyListResponse(
        items=[PolicyResponse(**item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )

# ── Create Policy ────────────────────────────────────────────────────────────

@router.post("", response_model=PolicyResponse, status_code=201)
async def create_policy_endpoint(
    request: PolicyCreateRequest,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Create a new policy."""
    _require_policy_role(user)
    tenant_id = str(user.tenant_id)
    user_id = str(user.user_id)
    await conn.set_tenant(tenant_id)

    # Validate the definition
    validation = validate_policy_definition(request.definition.model_dump())
    if not validation.valid:
        raise HTTPException(status_code=422, detail={"errors": validation.errors})

    try:
        result = await create_policy(conn, tenant_id, user_id, request)
    except Exception as e:
        error_msg = str(e)
        if "uq_policies_tenant_name" in error_msg:
            raise HTTPException(
                status_code=409,
                detail=f"Policy with name '{request.name}' already exists",
            )
        raise HTTPException(status_code=500, detail="Failed to create policy")

    return PolicyResponse(**result)

# ── Get Policy ───────────────────────────────────────────────────────────────

@router.get("/{policy_id}", response_model=PolicyResponse)
async def get_policy_endpoint(
    policy_id: uuid.UUID,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Get a policy by ID."""
    tenant_id = str(user.tenant_id)
    await conn.set_tenant(tenant_id)
    result = await get_policy(conn, tenant_id, str(policy_id))
    if result is None:
        raise HTTPException(status_code=404, detail="Policy not found")
    return PolicyResponse(**result)

# ── Update Policy ────────────────────────────────────────────────────────────

@router.put("/{policy_id}", response_model=PolicyResponse)
async def update_policy_endpoint(
    policy_id: uuid.UUID,
    request: PolicyUpdateRequest,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Update a policy (creates a new version)."""
    _require_policy_role(user)
    tenant_id = str(user.tenant_id)
    user_id = str(user.user_id)
    await conn.set_tenant(tenant_id)

    # Validate definition if provided
    if request.definition is not None:
        validation = validate_policy_definition(request.definition.model_dump())
        if not validation.valid:
            raise HTTPException(status_code=422, detail={"errors": validation.errors})

    result = await update_policy(conn, tenant_id, str(policy_id), user_id, request)
    if result is None:
        raise HTTPException(status_code=404, detail="Policy not found")

    return PolicyResponse(**result)

# ── Delete Policy ────────────────────────────────────────────────────────────

@router.delete("/{policy_id}", status_code=204)
async def delete_policy_endpoint(
    policy_id: uuid.UUID,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Soft-delete a policy (admin-recoverable)."""
    _require_policy_role(user)
    tenant_id = str(user.tenant_id)
    user_id = str(user.user_id)
    await conn.set_tenant(tenant_id)

    deleted = await delete_policy(conn, tenant_id, str(policy_id), user_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Policy not found")

# ── Validate Policy ──────────────────────────────────────────────────────────

@router.post("/validate", response_model=PolicyValidationResult)
async def validate_policy_endpoint(
    request: PolicyValidateRequest,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
):
    """Validate a policy definition without saving."""
    _require_policy_role(user)

    if request.yaml_content:
        parsed, errors = parse_yaml_safe(request.yaml_content)
        if errors:
            return PolicyValidationResult(valid=False, errors=errors)
        if parsed is None:
            return PolicyValidationResult(valid=False, errors=["Failed to parse YAML"])
        return validate_policy_definition(parsed)

    if request.json_content:
        return validate_policy_definition(request.json_content.model_dump())

    raise HTTPException(
        status_code=400,
        detail="Provide either yaml_content or json_content",
    )

# ── Apply Policy ─────────────────────────────────────────────────────────────

@router.post("/{policy_id}/apply")
async def apply_policy_endpoint(
    policy_id: uuid.UUID,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Apply policy to matching agents immediately."""
    _require_policy_role(user)
    tenant_id = str(user.tenant_id)
    await conn.set_tenant(tenant_id)

    policy = await get_policy(conn, tenant_id, str(policy_id))
    if policy is None:
        raise HTTPException(status_code=404, detail="Policy not found")

    if not policy["enabled"]:
        raise HTTPException(status_code=400, detail="Cannot apply disabled policy")

    # The actual application happens in the policy engine consumer.
    # Here we just validate and signal that the policy should be applied.
    definition = policy["definition"]
    rules_count = len(definition.get("rules", []))
    scope_tags = policy["scope_agent_tags"]
    scope_frameworks = policy["scope_frameworks"]

    logger.info(
        "policy_apply_requested",
        policy_id=str(policy_id),
        rules=rules_count,
        scope_tags=scope_tags,
        scope_frameworks=scope_frameworks,
    )

    return {
        "status": "applied",
        "policy_id": str(policy_id),
        "policy_name": policy["name"],
        "rules_count": rules_count,
        "scope": {
            "agent_tags": scope_tags,
            "frameworks": scope_frameworks,
        },
    }

# ── Version History ──────────────────────────────────────────────────────────

@router.get("/{policy_id}/versions", response_model=list[PolicyVersionResponse])
async def get_versions_endpoint(
    policy_id: uuid.UUID,
    user: Annotated[CurrentUser, Depends(get_current_active_user)],
    conn=Depends(get_raw_db),
):
    """Get version history for a policy."""
    tenant_id = str(user.tenant_id)
    await conn.set_tenant(tenant_id)

    # Verify policy exists
    policy = await get_policy(conn, tenant_id, str(policy_id))
    if policy is None:
        raise HTTPException(status_code=404, detail="Policy not found")

    versions = await get_policy_versions(conn, tenant_id, str(policy_id))
    return [PolicyVersionResponse(**v) for v in versions]
